Userland/find.cpp

```diff
@@ -0,0 +1,466 @@
+    }
+}
+
+OwnPtr<Command> parse_complex_command(char* argv[])
+{
+    auto command = parse_simple_command(argv);
+
+    while (command && argv[optind] && argv[optind + 1]) {
+        StringView arg = argv[++optind];
+
+        enum { And, Or } binary_operation = And;
+
+        if (arg == "-a") {
+            optind++;
+            binary_operation = And;
+        } else if (arg == "-o") {
+            optind++;
+            binary_operation = Or;
+        } else if (arg == ")") {
+            // Ooops, looked too far.
+            optind--;
+            return command;
+        } else {
+            // Juxtaposition is an And too, and there's nothing to skip.
+            binary_operation = And;
+        }
+
+        auto rhs = parse_complex_command(argv);
+        if (!rhs)
+            fatal_error("Missing right-hand side");
+
+        if (binary_operation == And)
+            command = make<AndCommand>(command.release_nonnull(), rhs.release_nonnull());
+        else
+            command = make<OrCommand>(command.release_nonnull(), rhs.release_nonnull());
+    }
+
+    return command;
+}
+
+NonnullOwnPtr<Command> parse_all_commands(char* argv[])
+{
+    auto command = parse_complex_command(argv);
+
+    if (g_have_seen_action_command) {
+        ASSERT(command);
+        return command.release_nonnull();
+    }
+
+    if (!command) {
+        return make<PrintCommand>();
+    }
+
+    return make<AndCommand>(command.release_nonnull(), make<PrintCommand>());
+}
+
+const char* parse_options(int argc, char* argv[])
+{
+    // Sadly, we can't use Core::ArgsParser, because find accepts arguments in
+    // an extremely unusual format. We're going to try to use getopt(), though.
+    opterr = 0;
+    while (true) {
+        int opt = getopt(argc, argv, "+L");
+        switch (opt) {
+        case -1: {
+            // No more options.
+            StringView arg = argv[optind];
+            if (!arg.is_null() && !arg.starts_with('-')) {
+                // It's our root path!
+                return argv[optind++];
+            } else {
+                // It's a part of the script, and our root path is the current
+                // directory by default.
+                return ".";
+            }
+        }
+        case '?':
+            // Some error. Most likely, it's getopt() getting confused about
+            // what it thought was an option, but is actually a command. Return
+            // the default path, and hope the command parsing logic deals with
+            // this.
+            return ".";
+        case 'L':
+            g_follow_symlinks = true;
+            break;
+        default:
+            ASSERT_NOT_REACHED();
+        }
+    }
+}
+
+void walk_tree(const char* root_path, Command& command)
+{
+    command.evaluate(root_path);
+
+    Core::DirIterator dir_iterator(root_path, Core::DirIterator::SkipParentAndBaseDir);
+    if (dir_iterator.has_error() && dir_iterator.error() == ENOTDIR)
+        return;
```

Comment:
Maybe say something?
I think this is about `find some-file -whatever`?

my host's `find` runs the commands on the single entry.

Reply:
It's not just about the top-level file; any time that we descend into the tree and find a file, we don't descend further (because you cannot descend into a file).

Reply:
Ah, the command is evaluated above.
Gotcha.